Sources/NIOHTTP2/HPACKHeaders+Validation.swift

```diff
@@ -0,0 +1,159 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import NIOHPACK
+
+extension HPACKHeaders {
+    /// Checks that a given HPACKHeaders block is a valid request header block, meeting all of the constraints of RFC 7540.
+    ///
+    /// If the header block is not valid, throws an error.
+    internal func validateRequestBlock() throws {
+        return try RequestBlockValidator.validateBlock(self)
+    }
+
+    /// Checks that a given HPACKHeaders block is a valid response header block, meeting all of the constraints of RFC 7540.
+    ///
+    /// If the header block is not valid, throws an error.
+    internal func validateResponseBlock() throws {
+        return try ResponseBlockValidator.validateBlock(self)
+    }
+
+    /// Checks that a given HPACKHeaders block is a valid trailer block, meeting all of the constraints of RFC 7540.
+    ///
+    /// If the header block is not valid, throws an error.
+    internal func validateTrailersBlock() throws {
+        return try TrailersValidator.validateBlock(self)
+    }
+}
+
+
+/// A `HeaderBlockValidator` is an object that can confirm that a HPACK block meets certain constraints.
+fileprivate protocol HeaderBlockValidator {
+    init()
+
+    mutating func validateNextField(name: HeaderFieldName, value: String) throws
+}
+
+extension HeaderBlockValidator {
+    /// Validates that a header block meets the requirements of this `HeaderBlockValidator`.
+    fileprivate static func validateBlock(_ block: HPACKHeaders) throws {
+        var validator = Self()
+        for (name, value, _) in block {
+            let fieldName = try HeaderFieldName(name)
+            try validator.validateNextField(name: fieldName, value: value)
+        }
+    }
+}
+
+/// An object that can be used to validate if a given header block is a valid request header block.
+fileprivate struct RequestBlockValidator { }
+
+extension RequestBlockValidator: HeaderBlockValidator {
+    fileprivate mutating func validateNextField(name: HeaderFieldName, value: String) throws {
+        return
```

Comment:
just interested... why we need a return here (+ in other "empty" functions)?

Reply:
We don't, they could have empty bodies.

Reply:
To clarify: I'm going to be extending these functions as I deal with other parts of RFC 7540 § 8.1.2. For now they are just spaces where code will go in future.